feat: Report missing before/after calls on undo corruption

Christopher-Chianelli opened the pull request.

Christopher-Chianelli committed:
feat: Report missing before/after calls on undo corruption

triceo reviewed (changes requested):
Right now, I think it's safe to assume that your needs for the current task will be evolving.

Let's wait until you've finished your work and then go through this code, figure out which parts we need, and what is the best way to package and deliver them.

Christopher-Chianelli committed:
fix: Add beforeListChangedEvent to before events

Christopher-Chianelli committed:
feat: Create a custom exception with access to the before, after and undo solutions

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/score/director/AbstractScoreDirector.java:
```diff
@@ -63,7 +67,18 @@ public abstract class AbstractScoreDirector<Solution_, Score_ extends Score<Scor
     protected Solution_ workingSolution;
     protected Integer workingInitScore = null;
-    private ShadowVariablesAssert beforeMoveSnapshot;
+    private AllVariablesAssert<Solution_> beforeMoveSnapshot;
+    private Solution_ beforeMoveSolution;
+    private AllVariablesAssert<Solution_> afterMoveSnapshot;
+    private Solution_ afterMoveSolution;
+    private final List<Pair<VariableDescriptor<Solution_>, Object>> beforeVariableChangedForwardEvents = new ArrayList<>();
+    private final List<Pair<VariableDescriptor<Solution_>, Object>> afterVariableChangedForwardEvents = new ArrayList<>();
+    private final List<Pair<VariableDescriptor<Solution_>, Object>> beforeVariableChangedUndoEvents = new ArrayList<>();
+    private final List<Pair<VariableDescriptor<Solution_>, Object>> afterVariableChangedUndoEvents = new ArrayList<>();
+
+    // These points to either ...ForwardEvents or ...UndoEvents, or null if not in an assert mode
+    private List<Pair<VariableDescriptor<Solution_>, Object>> beforeVariableChangedEvents = null;
+    private List<Pair<VariableDescriptor<Solution_>, Object>> afterVariableChangedEvents = null;
```
This complicates the score director far too much.

Let's think of a way of encapsulating it in a single type. All of this work should happen within that type, with the interactions between `ScoreDirector` and the other type being kept at a minimum;  thinking out loud here, perhaps some event-based system could do.

Once we have this in a separate class independent of the score director, we can test that it behaves properly in all cases. (Environment modes are largely untested, and with the code becoming really complicated, we need to change that.)

triceo commented on core/core-impl/src/test/java/ai/timefold/solver/core/config/solver/EnvironmentModeTest.java:
```diff
@@ -147,15 +155,23 @@ void corruptedUndoShadowVariableListener(EnvironmentMode environmentMode) {
 
         switch (environmentMode) {
             case FULL_ASSERT:
-            case FAST_ASSERT:
```
We need a name for the new environment mode, something more descriptive than `SUPER_FULL_ASSERT`. Or maybe not?

Your original numbers suggest that this is not *that much* slower than `FULL_ASSERT`. If you can first fix the score calculation numbers to give us *correct* data, we can assess those speeds again and if we're talking about a drop of a couple percent, we really don't need a whole new env mode for this.

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/score/director/AbstractScoreDirector.java:
```diff
@@ -644,53 +693,166 @@ public void assertExpectedUndoMoveScore(Move<Solution_> move, Score_ beforeMoveS
+        out.append(":\n");
+        int violationCount = 0;
+        for (var changedVariable : changedVariables) {
+            var expectedSnapshotVariable = expectedSnapshot.getVariableSnapshot(changedVariable);
+            var actualSnapshotVariable = actualSnapshot.getVariableSnapshot(changedVariable);
+            out.append("    ");
+            out.append(expectedSnapshotVariable.getVariableDescriptor().getSimpleEntityAndVariableName());
+            out.append(" (");
+            out.append(expectedSnapshotVariable.getEntity());
+            out.append(") expected (");
+            out.append(expectedSnapshotVariable.getValue());
+            out.append(") actual (");
+            out.append(actualSnapshotVariable.getValue());
+            out.append(")\n");
+            violationCount++;
+            if (violationCount >= limit) {
+                return;
+            }
```
Same comment on readable strings.

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/score/director/AbstractScoreDirector.java:
```diff
@@ -644,53 +693,166 @@ public void assertExpectedUndoMoveScore(Move<Solution_> move, Score_ beforeMoveS
-            throw new IllegalStateException("UndoMove corruption (" + scoreDifference
+            throw new UndoScoreCorruptionException("UndoMove corruption (" + scoreDifference
                     + "): the beforeMoveScore (" + beforeMoveScore + ") is not the undoScore (" + undoScore
                     + ") which is the uncorruptedScore (" + undoScore + ") of the workingSolution.\n"
-                    + differentShadowVariables
+                    + differentVariables
                     + "  1) Enable EnvironmentMode " + EnvironmentMode.FULL_ASSERT
                     + " (if you haven't already) to fail-faster in case there's a score corruption or variable listener corruption.\n"
                     + "  2) Check the Move.createUndoMove(...) method of the moveClass (" + move.getClass() + ")."
                     + " The move (" + move + ") might have a corrupted undoMove (" + undoMoveText + ").\n"
                     + "  3) Check your custom " + VariableListener.class.getSimpleName() + "s (if you have any)"
                     + " for shadow variables that are used by score constraints that could cause"
-                    + " the scoreDifference (" + scoreDifference + ").");
+                    + " the scoreDifference (" + scoreDifference + ").",
```
IMO we can also do a lot here to make the string much more readable, raw strings were made for this.

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/score/director/AbstractScoreDirector.java:
```diff
@@ -644,53 +693,166 @@ public void assertExpectedUndoMoveScore(Move<Solution_> move, Score_ beforeMoveS
-                    + "\n";
-        } else {
-            return "Shadow variables agrees with from scratch calculations before and after undo.\n";
+    private void appendVariableChangedEventViolations(StringBuilder out, String prefix,
```
Let's make this a function without side effects; return `String`, don't accept `out`.

Christopher-Chianelli committed:
chore: Move corruption analysis tracking code out of ScoreDirector

Christopher-Chianelli committed:
test: Add tests for SolutionTracker and friends

Christopher-Chianelli committed:
chore: add new enviroment mode to enable tracking

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/domain/variable/listener/support/violation/AllVariablesAssert.java:
```diff
@@ -0,0 +1,54 @@
+ * Serves for detecting undo move corruption. When a snapshot is created, it records the state of all variables (genuine and
+ * shadow) for all entities.
+ */
+public class AllVariablesAssert<Solution_> {
```
Also, is this an `Assert` if it doesn't actually assert anything? I'd call it `VariableSnapshotTotal` or something like that.

Christopher-Chianelli replied:
Counterpoint: `ShadowVariableAssert`.

triceo replied:
Good point. That said, if we're not already fixing mistakes of the past, at least we can not make new ones. WDYT?

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/domain/variable/listener/support/violation/VariableSnapshot.java:
```diff
@@ -0,0 +1,48 @@
+        return variableId;
+    }
+
+    public boolean isDifferentFrom(VariableSnapshot<Solution_> other) {
+        return !Objects.equals(value, other.value);
+    }
+}
```
Why not use `equals(...)` for this? Isn't that what it's for?

triceo replied:
In that case, yes, I agree on the exception. Failing fast is good, let's not be like a certain rule engine. ;-)

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/score/director/AbstractScoreDirector.java:
```diff
@@ -203,13 +211,16 @@ private void assertNonNullPlanningId(Object fact) {
     public Score_ doAndProcessMove(Move<Solution_> move, boolean assertMoveScoreFromScratch) {
-        if (assertMoveScoreFromScratch) {
-            beforeMoveSnapshot = ShadowVariablesAssert.takeSnapshot(getSolutionDescriptor(), workingSolution);
+        if (solutionTracker != null) {
```
I like the previous approach, where this condition was hidden behind a variable. The variable makes the purpose of the check obvious, while hiding what the check actually is.

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/score/director/AbstractScoreDirector.java:
```diff
@@ -607,7 +621,8 @@ public void assertShadowVariablesAreNotStale(Score_ expectedWorkingScore, Object
                     + "But all the shadow variable values are still the same, so this is impossible.\n"
-                    + "Maybe run with " + EnvironmentMode.FULL_ASSERT + " if you aren't already, to fail earlier.");
+                    + "Maybe run with " + EnvironmentMode.FULL_ASSERT_WITH_TRACKING
+                    + " if you aren't already, to fail earlier.");
```
Since we're touching this code, let's convert this nastiness to the new formatting pattern.

Remember: if you can, and it is not too disruptive, always leave the code in better shape than the one you found it in.

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/score/director/AbstractScoreDirector.java:
```diff
@@ -667,53 +682,41 @@ public void assertExpectedUndoMoveScore(Move<Solution_> move, Score_ beforeMoveS
             String scoreDifference = undoScore.subtract(beforeMoveScore).toShortString();
-
-            throw new IllegalStateException("UndoMove corruption (" + scoreDifference
+            String corruptionMessage = ("UndoMove corruption (" + scoreDifference
```
Dtto., formatting pattern.

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/score/director/AbstractScoreDirectorFactory.java:
```diff
@@ -68,6 +69,14 @@ public void setAssertClonedSolution(boolean assertClonedSolution) {
         this.assertClonedSolution = assertClonedSolution;
     }
 
+    public boolean isTrackingWorkingSolution() {
```
This deserves a Javadoc, pointing to someplace that describes what it will do. Perhaps a reference to the relevant `EnvironmentMode`?

triceo commented on core/core-impl/src/test/java/ai/timefold/solver/core/config/solver/EnvironmentModeTest.java:
```diff
@@ -88,17 +87,16 @@ void determinism(EnvironmentMode environmentMode) {
 
         switch (environmentMode) {
-            case NON_REPRODUCIBLE:
+            case NON_REPRODUCIBLE -> {
```
I like we're doing this!

Just please double check the entire codebase for switching over environment modes and apply this consistently. (If you haven't already.)

triceo commented on core/core-impl/src/test/java/ai/timefold/solver/core/impl/score/director/AbstractScoreDirectorSemanticsTest.java:
```diff
@@ -141,6 +141,8 @@ void constraintPresentEvenIfNoMatches() {
             scoreDirector.beforeVariableChanged(entity, "value");
             entity.setValue(null);
             scoreDirector.afterVariableChanged(entity, "value");
+            // FULL_ASSERT add variable listeners for tracking
+            scoreDirector.triggerVariableListeners();
             var score2 = scoreDirector.calculateScore();
```
This concerns me.
Are you saying we need to do something *new* in order to use this functionality? 
Why would this code need it, if it doesn't run any of the assert modes?

Christopher-Chianelli replied:
This code was introduced before `FULL_ASSERT` and `TRACKED_FULL_ASSERT` were seperated. `TRACKED_FULL_ASSERT` adds it own variable listeners, so it need the variable listeners to be trigerred to not throw an exception about missed before... and after... calls.

triceo replied:
Which effectively means it will *break* everything which didn't already do this.
`ScoreDirector` is public, its behavior is set in stone.
Therefore this is not acceptable.

triceo reviewed (changes requested):
I have added a whole bunch of comments, some to functionality, some to general code quality and cleanliness.

IMO we should discuss some general best practices we will follow throught every PR. We have some, but I feel like the list should be somewhat extended and modernized.

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/config/solver/EnvironmentMode.java:
```diff
@@ -19,6 +19,26 @@
 @XmlEnum
 public enum EnvironmentMode {
+    /**
+     * This mode turns on variable tracking and all assertions to fail-fast on a bug in a {@link Move} implementation, a
+     * constraint, the engine itself or something else at an appalling performance cost.
+     * <p>
+     * Because it tracks genuine and shadow variables, it is able to report precisely what variables caused the corruption and
+     * report any missed {@link ai.timefold.solver.core.api.domain.variable.VariableListener} events.
+     * <p>
+     * When undo score corruption is detected, it throws an
+     * {@link ai.timefold.solver.core.api.solver.exception.UndoScoreCorruptionException}, which can be used to inspect the
+     * working solution state before, after and after the undo of a {@link Move}.
+     * <p>
+     * This mode is reproducible (see {@link #REPRODUCIBLE} mode).
+     * <p>
+     * This mode is intrusive because it calls the {@link InnerScoreDirector#calculateScore()} more frequently than a non assert
+     * mode.
+     * <p>
+     * This mode is appallingly slow.
+     */
```
Besides the general comment on line breaks and formatting, I think maybe we're a bit more evolved now to use words such as "horrible" and "appaling" in user-facing documentation. :-) 
In this case, I suggest "This mode is by far the slowest of all the modes."

Also, somewhere in the body of the text, I would reference that it does everything that `FULL_ASSERT` does, plus X and Y. If you can reference an information that is already somewhere else, I prefer you do so - that way, you still have only a single point of truth for that information.

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/config/solver/EnvironmentMode.java:
```diff
@@ -19,6 +19,26 @@
+    FULL_ASSERT_WITH_TRACKING(true),
+
     /**
      * This mode turns on all assertions
```
This no longer turns on _all_ assertions.

Christopher-Chianelli replied:
It does turn on all assertions; `TRACKED_FULL_ASSERT` will only fail when `FULL_ASSERT` fails, the main difference is it tells you why it failed.

Christopher-Chianelli commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/domain/variable/listener/support/violation/VariableId.java:
```diff
@@ -0,0 +1,6 @@
+
+import ai.timefold.solver.core.impl.domain.variable.descriptor.VariableDescriptor;
+
+public record VariableId<Solution_>(VariableDescriptor<Solution_> variableDescriptor, Object entity) {
```
Well, the snapshots are taken from the same working solution, meaning the entities will be the same.

Christopher-Chianelli commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/domain/variable/listener/support/violation/ListVariableTracker.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    public TrackerDemand demand() {
+        return new TrackerDemand();
```
No two demands for these classes will ever be equal. It a hack in order to get the variable listeners set up, since they need a Demand to get put into the logic.

Christopher-Chianelli commented on core/core-impl/src/main/java/ai/timefold/solver/core/api/solver/exception/UndoScoreCorruptionException.java:
```diff
@@ -0,0 +1,100 @@
+package ai.timefold.solver.core.api.solver.exception;
```
This exception purely exists for the user; the internals do not use it. I myself has already used the exception to debug shadow variables. It allows you to do things like:

```
ObjectMapper mapper = ...;
try {
    ...
} catch (UndoScoreCorruptionException e) {
    Files.write("reproducer-solution.json", mapper.writeAsJson(e.getBeforeMoveSolution()));
    Files.write("reproducer-move.json", mapper.writeAsJson(e.getMove()));
}
```

to save the reproducer, or

```
try {
    ...
} catch (UndoScoreCorruptionException e) {
    // Debug point here
    e.reproduce();
}
```
To reproduce the exception and examine your own code when things went horribly wrong.
Strictly speaking, the reason why the solutions are cloned are purely for this exception; the snapshots has all the information we need for the exception message.

triceo replied:
Fair enough; my argument still stands though - having something in the public API means supporting it forever, and that means thinking a lot about all the possible implications of it.

If we put it outside of the public API, we can observe how people use it in real world, and _change it_ according to that usage. Think of this as the experimental package.

Please remove this from the public API.

Christopher-Chianelli replied:
We do need a place for experimental features we want people to use but understand they are not stable (yet). `impl` -> "Please do not use, unstable", `api` -> "Please use, stable", `?` -> "Please use, unstable"

triceo replied:
I agree in principle. In practice, it becomes more complex, as for example extensions to existing APIs (such as new methods on `SolutionManager`) make this more complicated.

Deserves a proper discussion, not part of this PR I'd argue.

Christopher-Chianelli commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/domain/variable/listener/support/violation/NormalVariableTracker.java:
```diff
@@ -0,0 +1,99 @@
+import ai.timefold.solver.core.impl.domain.variable.supply.Supply;
+import ai.timefold.solver.core.impl.domain.variable.supply.SupplyManager;
+
+public class NormalVariableTracker<Solution_>
```
Well this tracks both Genuine and Shadow variables (basically everything but List Variable)

triceo replied:
Then I'd argue we don't need the `Normal` prefix, and we do need a descriptive Javadoc.

Christopher-Chianelli committed:
chore: Address review comments

triceo commented:
@Christopher-Chianelli One thing I totally forgot - we should add this to the docs. AFAIK there is a whole section on assert modes, the new one should show up there too, with an explanation of what it's for, ideally with an example. Please do that, and then I'll merge it.

Christopher-Chianelli committed:
docs: Add section on TRACED_FULL_ASSERT

Christopher-Chianelli committed:
fix: Calculate before and undo from scratch seperately

The from scratch for before should use the before's genuine variables,
not the undo's genuine variables.

Christopher-Chianelli committed:
fix: Do not avoid genuine entities when resetting shadow variables

Genuine entities can also have shadow variables.

triceo commented on docs/src/modules/ROOT/pages/using-timefold-solver/running-the-solver.adoc:
```diff
@@ -102,6 +102,59 @@ This mode is reproducible (see the reproducible mode). It is also intrusive beca
+[#environmentModeTracedFullAssert]
+=== `TRACED_FULL_ASSERT`
+
+The TRACED_FULL_ASSERT mode turns on all the <<environmentModeFullAssert, FULL_ASSERT>> assertions and additionally track changes to the working solution.
```
```suggestion
The TRACED_FULL_ASSERT mode turns on all the <<environmentModeFullAssert, FULL_ASSERT>> assertions and additionally tracks changes to the working solution.
```

triceo commented on docs/src/modules/ROOT/pages/using-timefold-solver/running-the-solver.adoc:
```diff
@@ -102,6 +102,59 @@ This mode is reproducible (see the reproducible mode). It is also intrusive beca
+The TRACED_FULL_ASSERT mode turns on all the <<environmentModeFullAssert, FULL_ASSERT>> assertions and additionally track changes to the working solution.
+This allows the solver to report exactly what variables were corrupted and what variable listener events are missing.
+
+For instance, when this incorrect move is evaluated
```
```suggestion
Consider the following incorrectly implemented move:
```

triceo commented on docs/src/modules/ROOT/pages/using-timefold-solver/running-the-solver.adoc:
````diff
@@ -102,6 +102,59 @@ This mode is reproducible (see the reproducible mode). It is also intrusive beca
+    }
+```
+
+the exception thrown by the solver will have the following in its message:
````
```suggestion
When it is evaluated by the solver, an exception will be thrown with the following in its message:
```

triceo commented on docs/src/modules/ROOT/pages/using-timefold-solver/running-the-solver.adoc:
```diff
@@ -102,6 +102,59 @@ This mode is reproducible (see the reproducible mode). It is also intrusive beca
+
+In particular, the solver will recalculate all shadow variables from scratch on the solution after the undo and then report:
+
+- Genuine and shadow variables that are different between before and undo.
```
```suggestion
- Genuine and shadow variables that are different between "before" and "undo".
```

triceo commented on docs/src/modules/ROOT/pages/using-timefold-solver/running-the-solver.adoc:
```diff
@@ -102,6 +102,59 @@ This mode is reproducible (see the reproducible mode). It is also intrusive beca
+- Variables that are different between from scratch and before.
+This is to detect if the solution was corrupted before the move was evaluated due to shadow variable corruption.
+
+- Variables that are different between from scratch and undo.
```
```suggestion
- Variables that are different between "from scratch" and "undo".
```

triceo commented on docs/src/modules/ROOT/pages/using-timefold-solver/running-the-solver.adoc:
```diff
@@ -102,6 +102,59 @@ This mode is reproducible (see the reproducible mode). It is also intrusive beca
+Any variable that changed between the after move solution and after undo move solution without either a
+`beforeVariableChanged` or `afterVariableChanged` would be reported here.
+
+This mode is reproducible (see the reproducible mode).
```
The reproducible mode deserves a direct reference. (Unless it doesn't have an ID.)

triceo reviewed (approved):
I added some suggestions for improving the clarity of the doc, otherwise LGTM.

Christopher-Chianelli committed:
chore: apply suggestions from code review

Co-authored-by: Lukáš Petrovický <lukas@petrovicky.net>

Christopher-Chianelli committed:
chore: Add link to reproducible mode in the EnvironmentMode docs

triceo committed:
Finishing touches

triceo committed:
Merge branch 'main' into undo-corruption-full-analysis
